sdk/database/dbplugin/v5/grpc_server.go

```diff
@@ -16,19 +18,91 @@ var _ proto.DatabaseServer = gRPCServer{}
 type gRPCServer struct {
 	proto.UnimplementedDatabaseServer
 
-	impl Database
+	factoryFunc func() (Database, error)
+	instances   map[string]Database
+	sync.RWMutex
+}
+
+func getMultiplexIDFromContext(ctx context.Context) (string, error) {
+	md, ok := metadata.FromIncomingContext(ctx)
+	if !ok {
+		return "", fmt.Errorf("missing plugin multiplexing metadata")
+	}
+
+	multiplexIDs := md[multiplexingCtxKey]
+	if len(multiplexIDs) != 1 {
+		return "", fmt.Errorf("unexpected number of IDs in metadata: (%d)", len(multiplexIDs))
+	}
+
+	multiplexID := multiplexIDs[0]
+	if multiplexID == "" {
+		return "", fmt.Errorf("empty multiplex ID in metadata")
+	}
+
+	return multiplexID, nil
+}
+
+func (g gRPCServer) getOrCreateDatabase(ctx context.Context) (Database, error) {
+	g.Lock()
+	defer g.Unlock()
+
+	id, err := getMultiplexIDFromContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	if db, ok := g.instances[id]; ok {
+		return db, nil
+	}
+
+	db, err := g.factoryFunc()
+	if err != nil {
+		return nil, err
+	}
+
+	g.instances[id] = db
+
+	return db, nil
+}
+
+func (g gRPCServer) getDatabase(ctx context.Context) (Database, error) {
+	g.Lock()
```

Comment:
Can this be an `RLock`?

Reply:
As it currently stands I don't think so since `getDatabase` is called by `Close()`. I think for Close we will want a Write lock so that connections to the database aren't closed during other operations (read, write, update).

Maybe it makes sense to have a separate getDatabase call that holds a write lock for Close.

Reply:
addressed in https://github.com/hashicorp/vault/pull/13734/commits/05b0db6d7b8cafca01493b19b143d4c4a8653f1a